### ISSUE TITLE
pipeline: effect warning displayed twice

### ISSUE THREAD

**Issue by InfiniteProductions** [mono/MonoGame#4752]
Each warning is displayed twice in pipeline log output as the example below (at least on Windows platform, with pipeline version 3.5.1.1679):

> NormalMap.fx(72,31): warning X3206: 'mul': implicit truncation of vector type
NormalMap.fx(72,19): warning X3206: implicit truncation of vector type
NormalMap.fx(73,32): warning X3206: 'mul': implicit truncation of vector type
NormalMap.fx(73,20): warning X3206: implicit truncation of vector type
NormalMap.fx(74,33): warning X3206: 'mul': implicit truncation of vector type
NormalMap.fx(74,21): warning X3206: implicit truncation of vector type
NormalMap.fx(83,87): warning X3081: comma expression used where a vector constructor may have been intended
NormalMap.fx(83,92): warning X3081: comma expression used where a vector constructor may have been intended
NormalMap.fx(83,12): warning X3206: implicit truncation of vector type
NormalMap.fx(95,26): warning X3206: 'mul': implicit truncation of vector type
NormalMap.fx(95,12): warning X3206: implicit truncation of vector type
NormalMap.fx(72,31): warning X3206: 'mul': implicit truncation of vector type
NormalMap.fx(72,19): warning X3206: implicit truncation of vector type
NormalMap.fx(73,32): warning X3206: 'mul': implicit truncation of vector type
NormalMap.fx(73,20): warning X3206: implicit truncation of vector type
NormalMap.fx(74,33): warning X3206: 'mul': implicit truncation of vector type
NormalMap.fx(74,21): warning X3206: implicit truncation of vector type
NormalMap.fx(83,87): warning X3081: comma expression used where a vector constructor may have been intended
NormalMap.fx(83,92): warning X3081: comma expression used where a vector constructor may have been intended
NormalMap.fx(83,12): warning X3206: implicit truncation of vector type
NormalMap.fx(95,26): warning X3206: 'mul': implicit truncation of vector type
NormalMap.fx(95,12): warning X3206: implicit truncation of vector type

**Comment by tomspilman** [comment]
Likely there isn't anything for us to do here...  warnings and errors like this come from the HLSL compiler and we just present what it reports.

Still can you show us what some of this code looks like?  For example what are lines 72-74?

**Comment by InfiniteProductions** [comment]
Sure, this comes from RB Whitaker NormalMap shader:

>    output.Normal = normalize(mul(input.Normal, WorldInverseTranspose));
    output.Tangent = normalize(mul(input.Tangent, WorldInverseTranspose));
    output.Binormal = normalize(mul(input.Binormal, WorldInverseTranspose));

What can be done is parsing the result and while reading the output, discarding (with the setting to enable/disable it) each line which has already been processed (quite easy with some regexps and a dictionnary, I will eventually look at it)

**Comment by tomspilman** [comment]
> What can be done is parsing the result and while reading the output, discarding 

That sounds really error prone and dangerous.  If Microsoft's HLSL compiler or any other shader compiler returns results like this the only fix is in the compiler itself.  We can't just assume two errors for the same line means one of them is invalid.

**Comment by InfiniteProductions** [comment]
Of course, but assuming the two numbers are line and character position resp., + adding warning code and description shows obvious dupes in the example output.

**Comment by tomspilman** [comment]
You know this could just be an issue with redirecting console output.  I will investigate this for the 3.6 release.

**Comment by nkast** [comment]
I located the bug here:
https://github.com/MonoGame/MonoGame/blob/78e193e288672361260d82136ef2622a590c8e55/Tools/2MGFX/EffectObject.cs#L688-L698

It calls the compiler twice, once for  pixel shader and once for vertex. 
The second time it appends the same errors into `errorsAndWarnings`.
